infrastructure/src/main/java/org/corfudb/infrastructure/DynamicTriggerPolicy.java

```diff
@@ -0,0 +1,118 @@
+    private long lastCompactionCycleStartTS = 0;
+    private long lastAddressSpaceSizeAfterTrim = 0;
+
+    private CorfuRuntime corfuRuntime;
+    private Logger syslog;
+
+    public DynamicTriggerPolicy() {
+        lastCompactionCycleStartTS = System.currentTimeMillis();
+        lastAddressSpaceSizeAfterTrim = 0;
+        syslog = LoggerFactory.getLogger("SYSLOG");
+    }
+
+    @Override
+    public void markCompactionCycleStart() {
+        this.lastCompactionCycleStartTS = System.currentTimeMillis();
+    }
+
+    @Override
+    public long getLastCompactionCycleStartTime() {
+        return this.lastCompactionCycleStartTS;
+    }
+
+    @Override
+    public void markTrimComplete() {
+        this.lastAddressSpaceSizeAfterTrim = getCurrentAddressSpaceSize();
+    }
+
+    private long getCurrentAddressSpaceSize() {
+        long currentAddressSpaceSize = 0;
+        final int maxRetries = 10;
+        for (int retry = 0; retry < maxRetries; retry++) {
+            try {
+                currentAddressSpaceSize = corfuRuntime.getAddressSpaceView().getLogTail()
+                        - corfuRuntime.getAddressSpaceView().getTrimMark().getSequence();
+            } catch (Exception e) {
+                syslog.warn("getCurrentAddressSpaceSize hit exception {}. StackTrace {}", e, e.getStackTrace());
+            }
+        }
+        return currentAddressSpaceSize;
+    }
+
+    private boolean shouldForceTrigger() {
+        CorfuStore corfuStore = new CorfuStore(corfuRuntime);
+        try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
+            RpcCommon.TokenMsg upgradeToken = (RpcCommon.TokenMsg) txn.getRecord(DistributedCompactor.CHECKPOINT,
+                    DistributedCompactor.UPGRADE_KEY).getPayload();
+            txn.commit();
+            if (upgradeToken != null) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * 1. if ((currentTime - lastCompactionCycleStart) > minTimeBetweenCompactionStarts)
+     *        if (lastAddressSpaceSizeOnTrim == 0)
+     *            return true // no record of previous trim & safe trim period elapsed
+     *            ... once trim happens, we record the lastAddressSpaceSizeOnTrim
+     *            ... once checkpoint starts we record the lastCompactionCycleStartTS
+     * 2. if ((currentTime - lastCompactionCycleStart) > maxTimeBetweenCompactionStarts)
+     *
+     * @param interval - ignored for now
+     * @return true if compaction cycle should run, false otherwise
+     */
+    @Override
+    public boolean shouldTrigger(long interval) {
+
+        if (shouldForceTrigger()) {
+            syslog.info("Force triggering compaction");
+            return true;
+        }
+
+        final long currentTime = System.currentTimeMillis();
+        final long timeSinceLastCycleMillis = currentTime - lastCompactionCycleStartTS;
+        if (timeSinceLastCycleMillis > interval && lastAddressSpaceSizeAfterTrim == 0) {
```

Comment:
what does 0 mean?

Reply:
(lastAddressSpaceSizeAfterTrim == 0) could also mean that there are no writes to trim.  In that case, we shouldnt trim, right?

Reply:
The TriggerPolicy needs to be revisited after the scale test. And lastAddressSpaceSizeAfterTrim is not required rn. Thanks for the catch, will remove it.